File: blackbox/a2a-discussion-v0.1/runner.mjs
/**
 * Discussion derivation conformance runner v0.1
 *
 * Exercises the discussion-derivation corpus against the TypeScript reference
 * implementation (via dist/discussion.js). Pure offline — no server, no I/O
 * beyond reading the corpus file and the built module.
 *
 * Usage:
 *   node blackbox/a2a-discussion-v0.1/runner.mjs           # run corpus
 *   node blackbox/a2a-discussion-v0.1/runner.mjs --self     # self-test
 */
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { deriveDiscussion, parseEnvelope, parseReceiptAction } from "../../dist/discussion.js";

const ROOT = dirname(fileURLToPath(import.meta.url));
const CORPUS_PATH = join(ROOT, "corpus/v0.1/cases.json");
const PROFILE = "meshfleet.a2a.discussion-derivation.v0.1";

function encodeExpected(value) {
  if (value === null || value === undefined) return "null";
  if (typeof value === "boolean" || typeof value === "number") return JSON.stringify(value);
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(encodeExpected).join(",")}]`;
  return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${encodeExpected(value[key])}`).join(",")}}`;
}

function assertEqual(actual, expected, label) {
  if (encodeExpected(actual) !== encodeExpected(expected)) {
    throw new Error(`${label}: mismatch\n  actual:   ${JSON.stringify(actual)}\n  expected: ${JSON.stringify(expected)}`);
  }
}

function project(derived) {
  return {
    status: derived.status,
    turns_used: derived.turns_used,
    turns_remaining: derived.turns_remaining,
    transcript_length: derived.transcript.length,
    attempts_count: derived.attempts.length,
    integrity_finding_codes: derived.integrity_findings.map(f => f.code).sort(),
  };
}

function evaluateCase(item) {
  const input = item.input;

  if (input.action !== undefined) {
    const parsed = parseReceiptAction(input.action);
    if (parsed === null) {
      assertEqual(null, item.expected_output, item.id);
    } else {
      assertEqual(
        { kind: parsed.kind, state: parsed.state, turn: parsed.turn, attempt_id: parsed.attempt_id },
        item.expected_output,
        item.id,
      );
    }
    return;
  }

  if (input.payload !== undefined) {
    const parsed = parseEnvelope(input.payload);
    if (parsed === null) {
      assertEqual(null, item.expected_output, item.id);
    } else {
      assertEqual(
        {
          discussion_id: parsed.discussion_id, turn: parsed.turn, kind: parsed.kind,
          close: parsed.close, has_policy: parsed.policy !== undefined,
        },
        item.expected_output,
        item.id,
      );
    }
    return;
  }

  const derived = deriveDiscussion(input.discussion_id, input.messages, input.receipts, input.now);
  assertEqual(project(derived), item.expected_output, item.id);
}

function runCorpus() {
  const corpus = JSON.parse(readFileSync(CORPUS_PATH, "utf8"));
  if (corpus.profile !== PROFILE) throw new Error(`Profile mismatch: ${corpus.profile} !== ${PROFILE}`);
  for (const item of corpus.cases) evaluateCase(item);
  return { suite: "corpus", profile: PROFILE, cases: corpus.cases.length, passed: true };
}

function runSelf() {
  const result = deriveDiscussion("nonexistent", [], [], 1000);
  assertEqual(result.status, "invalid", "self-empty-invalid");
  assertEqual(result.integrity_findings.length > 0, true, "self-empty-has-findings");

  const parsed = parseReceiptAction("discussion.wake.completed.v1:2:att-x");
  assertEqual(parsed !== null, true, "self-parse-receipt");
  assertEqual(parsed.state, "completed", "self-receipt-state");

  const bad = parseReceiptAction("not.a.valid.action");
  assertEqual(bad, null, "self-bad-receipt");

  const moduleText = readFileSync(join(ROOT, "runner.mjs"), "utf8");
  if (!/PROFILE/.test(moduleText)) throw new Error("self-profile-constant");

  return { suite: "self", profile: PROFILE, cases: 4, passed: true };
}

const args = process.argv.slice(2);
if (args.length === 0 || (args.length === 1 && args[0] === "--corpus")) {
  process.stdout.write(JSON.stringify(runCorpus()));
} else if (args.length === 1 && args[0] === "--self") {
  process.stdout.write(JSON.stringify(runSelf()));
} else if (args.length === 1 && args[0] === "--hash-corpus") {
  process.stdout.write(createHash("sha256").update(readFileSync(CORPUS_PATH)).digest("hex"));
} else {
  throw new Error(`invalid arguments: ${args.join(" ")}`);
}

File: blackbox/a2a-discussion-v0.1/python/runner.py
#!/usr/bin/env python3
"""Independent Python runner for the discussion-derivation v0.1 witness."""

import argparse
import hashlib
import json
import re
import sys

from evaluator import derive_discussion, parse_envelope, parse_receipt_action


PROFILE = "meshfleet.a2a.discussion-derivation.v0.1"
MAX_DOCUMENT_BYTES = 12 * 1024 * 1024
MAX_JSON_DEPTH = 32


def _reject_constant(value):
    raise ValueError(f"nonstandard JSON constant: {value}")


def _object_without_duplicates(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            raise ValueError(f"duplicate JSON key: {key}")
        result[key] = value
    return result


def _check_depth(value):
    pending = [(value, 0)]
    while pending:
        current, depth = pending.pop()
        if depth > MAX_JSON_DEPTH:
            raise ValueError("JSON depth limit exceeded")
        if isinstance(current, dict):
            pending.extend((item, depth + 1) for item in current.values())
        elif isinstance(current, list):
            pending.extend((item, depth + 1) for item in current)


def _load_document(raw):
    if len(raw) > MAX_DOCUMENT_BYTES:
        raise ValueError("document size limit exceeded")
    document = json.loads(
        raw.decode("utf-8"),
        parse_constant=_reject_constant,
        object_pairs_hook=_object_without_duplicates,
    )
    _check_depth(document)
    return document


def _read_bounded(stream):
    raw = stream.read(MAX_DOCUMENT_BYTES + 1)
    if len(raw) > MAX_DOCUMENT_BYTES:
        raise ValueError("document size limit exceeded")
    return raw


def _validate_cases(cases):
    if not isinstance(cases, list) or not cases:
        raise ValueError("cases must be non-empty")
    case_ids = []
    seen_ids = set()
    for item in cases:
        if not isinstance(item, dict):
            raise ValueError("cases must be objects")
        case_id = item.get("id")
        if not isinstance(case_id, str) or not case_id or case_id in seen_ids:
            raise ValueError("case ids must be unique non-empty strings")
        seen_ids.add(case_id)
        case_ids.append(case_id)
    return case_ids


def _project_derived(derived, include_attempts=True):
    output = {
        "status": derived["status"],
        "turns_used": derived["turns_used"],
        "turns_remaining": derived["turns_remaining"],
        "transcript_length": len(derived["transcript"]),
        "integrity_finding_codes": sorted(
            finding["code"] for finding in derived["integrity_findings"]
        ),
    }
    if include_attempts:
        output["attempts_count"] = len(derived["attempts"])
    return output


def _evaluate_input(input_value, include_attempts=True):
    if "action" in input_value:
        parsed = parse_receipt_action(input_value["action"])
        if parsed is None:
            return None
        return {
            key: parsed[key]
            for key in ("kind", "state", "turn", "attempt_id")
        }

    if "payload" in input_value:
        parsed = parse_envelope(input_value["payload"])
        if parsed is None:
            return None
        return {
            "discussion_id": parsed["discussion_id"],
            "turn": parsed["turn"],
            "kind": parsed["kind"],
            "close": parsed["close"],
            "has_policy": "policy" in parsed,
        }

    derived = derive_discussion(
        input_value["discussion_id"],
        input_value["messages"],
        input_value["receipts"],
        input_value["now"],
    )
    return _project_derived(derived, include_attempts=include_attempts)


def _run_corpus(path, expected_sha256):
    with open(path, "rb") as handle:
        raw = _read_bounded(handle)
    actual_sha256 = hashlib.sha256(raw).hexdigest()
    if not re.fullmatch(r"[0-9a-f]{64}", expected_sha256 or ""):
        raise ValueError("expected sha256 must be 64 lowercase hex characters")
    if actual_sha256 != expected_sha256:
        raise ValueError(
            f"corpus sha256 mismatch: {actual_sha256} != {expected_sha256}"
        )
    corpus = _load_document(raw)
    if corpus.get("profile") != PROFILE:
        raise ValueError("unsupported profile")
    cases = corpus.get("cases")
    case_ids = _validate_cases(cases)

    failures = []
    for item in cases:
        actual = _evaluate_input(item["input"])
        expected = item["expected_output"]
        if actual != expected:
            failures.append(
                {"id": item["id"], "actual": actual, "expected": expected}
            )

    report = {
        "profile": PROFILE,
        "cases": len(cases),
        "corpus_sha256": actual_sha256,
        "case_ids": case_ids,
        "passed": not failures,
        "failures": failures,
    }
    print(json.dumps(report, sort_keys=True, separators=(",", ":")))
    return 0 if not failures else 1


def _evaluate_batch():
    document = _load_document(_read_bounded(sys.stdin.buffer))
    if document.get("profile") != PROFILE:
        raise ValueError("unsupported profile")
    cases = document.get("cases")
    _validate_cases(cases)

    outcomes = []
    seen_ids = set()
    for item in cases:
        case_id = item.get("id")
        if not isinstance(case_id, str) or not case_id or case_id in seen_ids:
            raise ValueError("case ids must be unique non-empty strings")
        seen_ids.add(case_id)
        outcomes.append(
            {
                "id": case_id,
                "output": _evaluate_input(
                    item["input"],
                    include_attempts=False,
                ),
            }
        )

    print(
        json.dumps(
            {"profile": PROFILE, "outcomes": outcomes},
            sort_keys=True,
            separators=(",", ":"),
        )
    )
    return 0


def main():
    parser = argparse.ArgumentParser()
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--corpus")
    group.add_argument("--evaluate-json", action="store_true")
    parser.add_argument("--expected-sha256")
    args = parser.parse_args()

    try:
        if args.corpus:
            return _run_corpus(args.corpus, args.expected_sha256)
        return _evaluate_batch()
    except Exception as error:
        print(
            json.dumps(
                {
                    "profile": PROFILE,
                    "passed": False,
                    "error": f"{type(error).__name__}: {error}",
                    "failures": [],
                },
                sort_keys=True,
                separators=(",", ":"),
            )
        )
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
